### PR TITLE
[Profiling] Support downsampling of generic events

### PR DESCRIPTION
**Pull request by danielmitterdorfer**
We have recently added a more generic interface that allows to gather stacktrace events from a user-provided index. However, we have always considered all samples as is which does not scale.

With this commit we introduce dynamic downsampling based on the random sampler aggregation so we provide only (roughly) the requested number of samples.

### CODE REVIEW

File: x-pack/plugin/profiling/src/internalClusterTest/java/org/elasticsearch/xpack/profiling/GetStackTracesActionIT.java [hunk 1]
```diff
@@ -80,6 +86,70 @@ public void testGetStackTracesFromAPMWithMatch() throws Exception {
...
+            null,
+            null
+        ) {
+            @Override
+            public boolean equals(Object o) {
+                return super.equals(o);
+            }
+
+            @Override
+            public int hashCode() {
```
Comment (danielmitterdorfer) [inline x-pack/plugin/profiling/src/internalClusterTest/java/org/elasticsearch/xpack/profiling/GetStackTracesActionIT.java:112]: I'm not fond of this approach but I've started an internal discussion how we can improve the situation when the random sampler aggregation is involved. Nevertheless, I did not want to block progress on the PR.

**Comment by danielmitterdorfer** [comment]
A CI check failed due to https://github.com/elastic/elasticsearch/issues/104776 which will be addressed by https://github.com/elastic/elasticsearch/pull/104780.

**Comment by danielmitterdorfer** [comment]
@elasticmachine merge upstream

File: x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/GetStackTracesRequest.java [hunk 2]
```diff
@@ -98,7 +98,7 @@ public void writeTo(StreamOutput out) {
         TransportAction.localOnly();
     }
 
-    public Integer getSampleSize() {
+    public int getSampleSize() {
         return sampleSize != null ? sampleSize : DEFAULT_SAMPLE_SIZE;
```
Comment (rockdaboot) [inline x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/GetStackTracesRequest.java:102]: IMO, we can make `sampleSize` an `int` and move the null check into the constructor.

File: x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java [hunk 3]
```diff
@@ -216,8 +218,42 @@ private void searchGenericEvents(
...
+            .setSize(0)
+            .setTrackTotalHits(true)
+            .setRequestCache(true)
+            .setPreference(String.valueOf(request.hashCode()))
+            .setQuery(request.getQuery())
+            .execute(ActionListener.wrap(searchResponse -> {
+                long sampleCount = searchResponse.getHits().getTotalHits().value;
+                int requestedSampleCount = request.getSampleSize();
+                // random sampler aggregation does not support sampling rates between 0.5 and 1.0 -> clamp to 1.0
+                if (sampleCount <= requestedSampleCount * 2L) {
```
Comment (rockdaboot) [inline x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java:232]: Having these values hard-coded isn't ideal. Can we somehow read these values out from the liable component? Or change that component that it automatically uses 1.0 when the samplingRate is in the unsupported area?

**Review by rockdaboot — APPROVED** [review]
Using the built-in random downsampling is a huge improvement :+1:.
Though it accesses the whole range of data in a random manner and may be not as efficient as expected when dealing with large amounts of data (we had a latency issue with profiling data).

The other comments are not meant as blockers.

File: x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java (hunk 3 above)
Comment (danielmitterdorfer) [reply x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java:232]: Yeah, I'm not a fan of hard-coded values either. However, in this case I believe it's appropriate. The fact that random sampler disallows probabilities between (0.5;1) is just [documented knowledge](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-random-sampler-aggregation.html) but is not retrievable from the code (i.e. hardcoded there as well):

https://github.com/elastic/elasticsearch/blob/72fe8b30d5bc69b69e3af0c0a9e6478f362a6f34/server/src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/random/RandomSamplerAggregationBuilder.java#L58-L60

Basically this code is here to adhere to the preconditions of the called code and I'd argue that having the constant embedded here is ok.

File: x-pack/plugin/profiling/src/internalClusterTest/java/org/elasticsearch/xpack/profiling/GetStackTracesActionIT.java (hunk 1 above)
Comment (danielmitterdorfer) [reply x-pack/plugin/profiling/src/internalClusterTest/java/org/elasticsearch/xpack/profiling/GetStackTracesActionIT.java:112]: Currently there is https://github.com/elastic/elasticsearch/pull/104830 open to address this. Once this is merged I'd open a small follow-up PR to simplify this test.

**Comment by danielmitterdorfer** [comment]
Thanks for the review!

>  Though it accesses the whole range of data in a random manner and may be not as efficient as expected when dealing with large amounts of data (we had a latency issue with profiling data).

Yes, that's understood and accepted for now. We do expect to take this operation up to 500ms based on internal benchmark results. While it's not great it will allow us gauge whether that is ok in practice without impacting storage requirements upfront.
